test/gradualizer_bin_tests.erl

```diff
@@ -0,0 +1,62 @@
+-module(gradualizer_bin_tests).
+
+-include_lib("eunit/include/eunit.hrl").
+
+%% Macro to convert type to abstract form
+-define(t(T), typelib:remove_pos(typelib:parse_type(??T))).
+
+%% Macro to convert expression to abstract form
+-define(e(E), merl:quote(??E)).
+
+-define(_assert_bin_type(T, E),
+        ?_assertEqual(?t(T), gradualizer_bin:compute_type(?e(E)))).
+
+compute_type_combined_test_() ->
+    [
+     ?_assert_bin_type(<<_:14, _:_*3>>, <<0:14, N/binary-unit:3>>),
+     ?_assert_bin_type(<<_:14, _:_*3>>, <<0:N/integer-unit:3, "ab":7/integer>>)
+    ].
+
+compute_type_utf_test_() ->
+    [
+     ?_assert_bin_type(<<_:24>>,    <<"abc"/utf8>>),
+     ?_assert_bin_type(<<_:48>>,    <<"abc"/utf16>>),
+     ?_assert_bin_type(<<_:96>>,    <<"abc"/utf32>>),
+     ?_assert_bin_type(<<_:_*8>>,   <<X/utf8>>),
+     ?_assert_bin_type(<<_:_*16>>,  <<X/utf16>>),
+     ?_assert_bin_type(<<_:32>>,    <<X/utf32>>)
+    ].
+
+compute_type_test_() ->
+    [
+     ?_assert_bin_type(<<_:8>>,     <<X>>),
+     ?_assert_bin_type(<<_:3>>,     <<X:3>>),
+     ?_assert_bin_type(<<_:_*1>>,   <<X:N>>),
+     ?_assert_bin_type(<<_:12>>,    <<X:3/unit:4>>),
+     ?_assert_bin_type(<<_:_*4>>,   <<X:N/unit:4>>)
+    ].
+
+compute_type_bitstring_test_() ->
+    [
+     ?_assert_bin_type(<<_:8>>,     <<X:1/binary>>),
+     ?_assert_bin_type(<<_:42>>,    <<X:7/binary-unit:6>>),
+     ?_assert_bin_type(<<_:_*8>>,   <<X/binary>>),
+     ?_assert_bin_type(<<_:_*2>>,   <<X/binary-unit:2>>),
+     ?_assert_bin_type(<<_:_*8>>,   <<X/bytes>>),
+     ?_assert_bin_type(<<_:16>>,    <<X:2/bytes>>),
+     ?_assert_bin_type(<<_:_*1>>,   <<X/bitstring>>),
+     ?_assert_bin_type(<<_:_*1>>,   <<X/bits>>)
+    ].
+
+compute_type_float_test_() ->
+    [
+     ?_assert_bin_type(<<_:64>>,            <<X/float>>),
+     ?_assert_bin_type(<<_:32>>,            <<X:32/float>>),
+     ?_assert_bin_type(<<_:32, _:_*32>>,    <<X:S/float>>),
+     ?_assert_bin_type(<<_:32, _:_*32>>,    <<X:S/float-unit:16>>),
+     ?_assert_bin_type(<<_:64>>,            <<X:S/float-unit:64>>),
+     ?_assert_bin_type(<<_:192>>,           <<"abc"/float>>),
```

Comment:
`<<"abc"/float>>` is not supported on OTP 19 and travis fails with it (but ok on OTP 20+)
good to know :), I did not know this syntax (as it turns out equivalent with `<<$a/float, $b/float, $c/float>>` and <<97/float, 98/float, 99/float>>`)

Reply:
Thanks for finding the cause of this! Any suggestion on how to skip these cases on OTP 19?

Reply:
It's a pretty useless syntax btw. Wonder why they added it. Once I suggested on erlang-questions to add `<<BoolExpr:1/boolean>>` but that was not popular at all. :-)

Reply:
- maybe using the `OTP_RELEASE` macro (and test only from OTP 21+)
- or as it fails at runtime (not compiler error), there could be a helper function which checks if this syntax throws a badarg

Reply:
I remember the `/boolean` issue :) the community couldn't decide if `0` should be true or false...

Reply:
> It's a pretty useless syntax btw. Wonder why they added it.

Some git archaeology suggests that it was added unintentionally in erlang/otp@a5fcd4f269.

Reply:
Impressive finding @UlfNorell! And Jose Valim is to blame :) I don't mind that this syntax exists, even if it is useless.